substrate/client/rpc-servers/src/middleware/rate_limit.rs

```diff
@@ -0,0 +1,80 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! RPC rate limiting middleware.
+
+use futures::future::{BoxFuture, FutureExt};
+use governor::{
+	clock::DefaultClock,
+	middleware::NoOpMiddleware,
+	state::{InMemoryState, NotKeyed},
+	Jitter,
+};
+use jsonrpsee::{server::middleware::rpc::RpcServiceT, types::Request, MethodResponse};
+use std::{num::NonZeroU32, sync::Arc, time::Duration};
+
+type RateLimitInner = governor::RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>;
+const MAX_JITTER_DELAY: Duration = Duration::from_millis(50);
+
+/// JSON-RPC rate limit middleware layer.
+#[derive(Debug, Clone)]
+pub struct RateLimitLayer(governor::Quota);
+
+impl RateLimitLayer {
+	/// Create new rate limit enforced per minute.
+	///
+	/// # Panics
+	///
+	/// Panics if n is zero.
+	pub fn per_minute(n: u32) -> Self {
+		Self(governor::Quota::per_minute(NonZeroU32::new(n).unwrap()))
```

Comment:
This can only happen if  `--rpc-rate-limit 0` which doesn't make any sense, I could refactor this check that n > 0 in the code or just disable the rate-limit for that case.

Any suggestions?

Reply:
Could the CLI arg take in an `Option<NonZeroU32>` instead of `Option<u32>` to prevent this from being a possibility?

Reply:
yepp, since no other CLI stuff is using NonZeroInteger types  I wanted to be consistent but worth breaking that.